// @vuetify/loader-shared/dist/styles/writeStyles.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.writeStyles = exports.cacheDir = void 0;
const promises_1 = require("fs/promises");
const find_cache_dir_1 = __importDefault(require("find-cache-dir"));
const path = __importStar(require("upath"));
exports.cacheDir = (0, find_cache_dir_1.default)({
    name: 'vuetify',
    create: true,
    thunk: true
});
function normalize(p) {
    p = path.normalize(p);
    return /^[a-z]:\//i.test(p) ? '/' + p : p;
}
function writeStyles(files) {
    return (0, promises_1.writeFile)((0, exports.cacheDir)('styles.scss'), ['vuetify/lib/styles/main.sass', ...files.values()].map(v => `@forward '${normalize(v)}';`).join('\n'), 'utf8');
}
exports.writeStyles = writeStyles;
